src/services/services.ts

```diff
@@ -6259,6 +6373,187 @@ namespace ts {
+                        getSymbolsForComponentTypes(<UnionOrIntersectionType>componentType, result);
+                    }
+                }
+                return result;
+            }
+
+            function getContainingTypeReference(node: Node): Node {
+                let topLevelTypeReference: Node = undefined;
+
+                while (node) {
+                    if (isTypeNode(node)) {
+                        topLevelTypeReference = node;
+                    }
+                    node = node.parent;
+                }
+
+                return topLevelTypeReference;
+            }
+
+            function getContainingClassHeritageClause(node: Node): HeritageClause {
+                if (node) {
+                    if (node.kind === SyntaxKind.ExpressionWithTypeArguments
+                        && node.parent.kind === SyntaxKind.HeritageClause
+                        && isClassLike(node.parent.parent)) {
+                        return <HeritageClause>node.parent;
+                    }
+
+                    else if (node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.PropertyAccessExpression) {
+                        return getContainingClassHeritageClause(node.parent);
+                    }
+                }
+                return undefined;
+            }
+
+            /**
+             * Returns true if this is an expression that could be used to implement an interface.
+             */
+            function isImplementationExpression(node: Expression): boolean {
+                // Unwrap parentheses
+                if (node.kind === SyntaxKind.ParenthesizedExpression) {
+                    return isImplementationExpression((<ParenthesizedExpression>node).expression);
+                }
+
+                return node.kind === SyntaxKind.ArrowFunction ||
+                    node.kind === SyntaxKind.FunctionExpression ||
+                    node.kind === SyntaxKind.ObjectLiteralExpression ||
+                    node.kind === SyntaxKind.ClassExpression ||
+                    node.kind === SyntaxKind.ArrayLiteralExpression;
+            }
+
+            /**
+             * Determines if the parent symbol occurs somewhere in the child's ancestry. If the parent symbol
+             * is an interface, determines if some ancestor of the child symbol extends or inherits from it.
+             * This also takes in a cache of previous results which makes this slightly more efficient and is
+             * necessary to avoid potential loops like so:
+             *     class A extends B { }
+             *     class B extends A { }
+             *
+             * @param child         A class or interface Symbol
+             * @param parent        Another class or interface Symbol
+             * @param cachedResults A map of symbol names to booleans indicating previous results
+             */
+            function inheritsFrom(child: Symbol, parent: Symbol, cachedResults: Map<boolean>): boolean {

```

Comment:
I was originally using the type checker, but it presented a bunch of false positives where another class or interface implementation structurally matched but did not explicitly inherit from the type in question. An example:
```typescript
class BaseFoo {
    hello() {}
}

class Foo extends BaseFoo {}

class SiblingToFoo extends BaseFoo {
    hello() {}
}

function someFunction (x: Foo) {
    x.hello // Go to implementation here
}
```
If we go to implementation on `x.hello`, the symbol that will be searched for in find-all-references is `BaseFoo.hello` which matches both the implementations of `hello` above. The only result the user actually wants is the implementation in `BaseFoo`, but the type checker doesn't really make the differentiation between `Foo` and `SiblingToFoo`. There is also an equivalent case for interfaces. Walking the AST enables us to only return explicit implementations.